tor-android-binary/src/main/java/org/torproject/jni/TorService.java

```diff
@@ -327,66 +325,65 @@ public void run() {
+                    }
+                    int result = runMain(); // run verify
+                    if (result != 0) {
+                        throw new IllegalArgumentException("Bad command flags: " + Arrays.toString(verifyLines));
+                    }
 
-                        controlPortThreadStarted = new CountDownLatch(1);
-                        controlPortThread.start();
-                        controlPortThreadStarted.await();
+                    controlPortThreadStarted = new CountDownLatch(1);
+                    controlPortThread.start();
+                    controlPortThreadStarted.await();
 
-                        String[] runLines = new String[lines.size() - 1];
-                        runLines[0] = "tor";
-                        for (int i = 2; i < lines.size(); i++) {
-                            runLines[i - 1] = lines.get(i);
-                        }
-                        if (!mainConfigurationSetCommandLine(runLines)) {
-                            throw new IllegalArgumentException("Setting command line failed: " + Arrays.toString(runLines));
-                        }
-                        if (!mainConfigurationSetupControlSocket()) {
-                            throw new IllegalStateException("Setting up ControlPort failed!");
-                        }
-                        if (runMain() != 0) {
-                            throw new IllegalStateException("Tor could not start!");
-                        }
-                    } finally {
-                        mainConfigurationFree();
-                        Log.i(TAG, "Releasing lock");
-                        runLock.unlock();
+                    String[] runLines = new String[lines.size() - 1];
+                    runLines[0] = "tor";
+                    for (int i = 2; i < lines.size(); i++) {
+                        runLines[i - 1] = lines.get(i);
+                    }
+                    if (!mainConfigurationSetCommandLine(runLines)) {
+                        throw new IllegalArgumentException("Setting command line failed: " + Arrays.toString(runLines));
+                    }
+                    if (!mainConfigurationSetupControlSocket()) {
+                        throw new IllegalStateException("Setting up ControlPort failed!");
+                    }
+                    if (runMain() != 0) {
+                        throw new IllegalStateException("Tor could not start!");
                     }
 
                 } catch (IllegalStateException | IllegalArgumentException | InterruptedException e) {
                     e.printStackTrace();
                     broadcastError(context, e);
+                } finally {
                     broadcastStatus(context, STATUS_STOPPING);
+                    mainConfigurationFree();
+                    Log.i(TAG, "Releasing lock");
+                    runLock.unlock();
```

Comment:
Maybe `shutdownTor(TorControlCommands.SIGNAL_SHUTDOWN);` should also be called here?

Reply:
tor_main() has returned at this point so I don't think it could have any effect.